src/AudioInputI2S.cpp

```diff
@@ -0,0 +1,215 @@
+  } 
+#else
+  (void) use_apll;
+  if (!running) {
+    i2s_begin();
+  }
+#endif
+  running = true;
+  bps = 32;
+  channels = 1;
+  gain_shift = 0;
+}
+
+AudioInputI2S::~AudioInputI2S()
+{
+#ifdef ESP32
+  if (running) {
+    Serial.printf("UNINSTALL I2S\n");
+    i2s_driver_uninstall((i2s_port_t)portNo); //stop & destroy i2s driver
+  }
+#else
+  if (running) i2s_end();
+#endif
+running = false;
+free(buff);
+}
+
+uint32_t AudioInputI2S::GetSample()
+{
+  if (!running) return 0;
+
+  uint32_t sampleIn=0;
+#ifdef ESP32
+  i2s_pop_sample((i2s_port_t)portNo, (char*)&sampleIn, portMAX_DELAY);
+  return sampleIn>>(14+gain_shift);
+#else
+  i2s_read_sample(&lastSample[LEFTCHANNEL], &lastSample[RIGHTCHANNEL], true);
+  return (lastSample[RIGHTCHANNEL] << 16) | (lastSample[LEFTCHANNEL] & 0xffff);
+#endif
+}
+
+uint32_t AudioInputI2S::read(void* data, size_t len_bytes)
+{
+  if (!running) return 0;
+  size_t bytes_read = 0;
+  //bytes_read = i2s_read_bytes((i2s_port_t)portNo, (char*)data, (size_t)len, portMAX_DELAY);
+#ifdef ESP32
+  esp_err_t err = i2s_read((i2s_port_t)portNo, data, len_bytes, &bytes_read, 0);
+#else
+  uint16_t* data16 = reinterpret_cast<uint16_t*>(data);
+  for (int i = 0; (i < len_bytes/(2*sizeof(int16_t))) && (i2s_rx_available()) > 0; i++) {
+    i2s_read_sample(&lastSample[LEFTCHANNEL], &lastSample[RIGHTCHANNEL], false);
+    data16[i+RIGHTCHANNEL] = lastSample[RIGHTCHANNEL];
+    data16[i+LEFTCHANNEL] = lastSample[LEFTCHANNEL];
+  }
+#endif
+  return bytes_read;
+}
+
+bool AudioInputI2S::SetPinout(int bclk, int wclk, int din)
+{
+#ifdef ESP32
+  i2s_pin_config_t pin_config = {
+    .bck_io_num = bclk,
+    .ws_io_num = wclk,
+    .data_out_num = I2S_PIN_NO_CHANGE,
+    .data_in_num = din
+  };
+  i2s_set_pin((i2s_port_t)portNo, &pin_config);
+  return true;
+#else
+  (void) bclk;
+  (void) wclk;
+  (void) din;
+  return false;
+#endif
+}
+
+bool AudioInputI2S::SetRate(int hz)
+{
+  // TODO - have a list of allowable rates from constructor, check them
+  this->hertz = hz;
+#ifdef ESP32
+  i2s_set_sample_rates((i2s_port_t)portNo, hz); 
+#else
+  i2s_set_rate(hz);
+#endif
+  return true;
+}
+
+bool AudioInputI2S::SetBitsPerSample(int bits)
+{
+  if ( (bits != 32) && (bits != 16) && (bits != 8) ) return false;
+  this->bps = bits;
+  return true;
+}
+
+bool AudioInputI2S::SetGain(float f)
+{
+  if (f < 0.33) this->gain_shift = 2;
+  else if (f < 0.66) this->gain_shift = 1;
+  else this->gain_shift = 0;
+  return true;
+}
+
+bool AudioInputI2S::stop() {
+  running = false;
+  output->stop();
+#ifdef ESP32
+  esp_err_t err = i2s_stop((i2s_port_t)portNo);
+  return (err == ESP_OK);
+#else
+  i2s_rxtx_begin(false, false);
+  return true;
+#endif
+}
+
+bool AudioInputI2S::isRunning() {
+  return running;
+}
+
+bool AudioInputI2S::loop() {
+  if (!running) return false;
+
+  uint32_t* buff32 = reinterpret_cast<uint32_t*>(buff);
+
+  while (validSamples) {
+    int32_t sample = buff32[curSample]>>(14+gain_shift);
```

Comment:
I think this code and below doesn't work for 16-bits or stereo mikes.  On the ESP8266 I only implemented 16-bit stereo to match the fixed output format on that system (and since my only testing was on a dual-mike setup from the RPI hat from GOOG).

Also, because the stream is a set of fixed-point values from -1 to 0.9999, I think you can get much finer grained amplification simply by integer multiply and shifting, like (for example) is done in the Amplify() method.